Rare crash attempting to auto-save but finding a non-empty directory

This issue has popped up fairly rarely since auto-save was implemented, too rare to be able to reproduce easily. Maybe a race condition of some sort, everybody loves those!

I've got a "Terasology" alert on pastebin so I catch some of these even when users haven't submitted a support request. Since I started I've seen this issue happen 2-3 times (over around 6 weeks). So I thought we should at least make an issue for it, perhaps find a way to improve the logging to get more details when it does happen.

Latest paste (expires in 29 days): http://pastebin.com/9vg8TXT7

It does have a sort of unusual module lineup. Looks kinda like JoshariasSurvival, but not exactly, also has WildAnimals + Pathfinding.

Maybe interesting to @flo although likely difficult to debug further without more logging / easier way to reproduce.

Snippet:

```
23:13:07.600 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: engine:0.54.5
23:13:07.600 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: Core:0.54.5
23:13:07.600 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: CoreSampleGameplay:0.54.5
23:13:07.600 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: CustomOreGen:0.1.1-SNAPSHOT
23:13:07.600 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: Durability:0.1.1-SNAPSHOT
23:13:07.600 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: EventualSkills:0.1.0-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: Fluid:0.1.1-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: InGameHelp:0.1.0-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: ItemRendering:0.1.2-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: Workstation:0.1.1-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: Machines:0.1.2-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: OreGeneration:0.1.1-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: SubstanceMatters:0.1.1-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: ManualLabor:0.1.1-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: ManualLaborEventualSkills:0.1.0-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: MarkovChains:1.5.1-SNAPSHOT
23:13:07.601 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: NameGenerator:0.4.1-SNAPSHOT
23:13:07.602 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: Pathfinding:0.1.2-SNAPSHOT
23:13:07.602 [main] INFO  o.t.e.m.loadProcesses.RegisterMods - Activating module: WildAnimals:0.1.1-SNAPSHOT

............

23:16:25.589 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
23:16:25.591 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
23:16:31.015 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
23:17:25.609 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
23:17:25.611 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
23:17:30.417 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
23:18:25.924 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
23:18:25.925 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
23:18:31.591 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
23:19:25.944 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
23:19:25.946 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
23:19:29.159 [Saving-0] ERROR o.t.p.internal.SaveTransaction - Save game creation failed
java.nio.file.DirectoryNotEmptyException: C:\Users\Marcel PC\Saved Games\Terasology\saves\Game 6\unmerged-changes
    at sun.nio.fs.WindowsFileSystemProvider.implDelete(Unknown Source) ~[na:1.8.0_71]
    at sun.nio.fs.AbstractFileSystemProvider.delete(Unknown Source) ~[na:1.8.0_71]
    at java.nio.file.Files.delete(Unknown Source) ~[na:1.8.0_71]
    at org.terasology.persistence.internal.SaveTransactionHelper$2.postVisitDirectory(SaveTransactionHelper.java:103) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransactionHelper$2.postVisitDirectory(SaveTransactionHelper.java:72) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at java.nio.file.Files.walkFileTree(Unknown Source) ~[na:1.8.0_71]
    at java.nio.file.Files.walkFileTree(Unknown Source) ~[na:1.8.0_71]
    at org.terasology.persistence.internal.SaveTransactionHelper.mergeChanges(SaveTransactionHelper.java:72) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransaction.mergeChanges(SaveTransaction.java:465) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransaction.run(SaveTransaction.java:139) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.utilities.concurrency.TaskProcessor.run(TaskProcessor.java:49) [engine-0.54.5.jar:62, master, 62, pre-alpha]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [na:1.8.0_71]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [na:1.8.0_71]
    at java.lang.Thread.run(Unknown Source) [na:1.8.0_71]
23:19:29.172 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.RuntimeException: Saving failed
    at org.terasology.persistence.internal.ReadWriteStorageManager.checkSaveTransactionAndClearUpIfItIsDone(ReadWriteStorageManager.java:158) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.ReadWriteStorageManager.update(ReadWriteStorageManager.java:383) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.engine.modes.StateIngame.update(StateIngame.java:184) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:414) ~[engine-0.54.5.jar:62, origin/master, 62]
    at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:369) ~[engine-0.54.5.jar:62, origin/master, 62]
    at org.terasology.engine.Terasology.main(Terasology.java:154) [Terasology.jar:62, origin/master, 62]
Caused by: java.nio.file.DirectoryNotEmptyException: C:\Users\Marcel PC\Saved Games\Terasology\saves\Game 6\unmerged-changes
    at sun.nio.fs.WindowsFileSystemProvider.implDelete(Unknown Source) ~[na:1.8.0_71]
    at sun.nio.fs.AbstractFileSystemProvider.delete(Unknown Source) ~[na:1.8.0_71]
    at java.nio.file.Files.delete(Unknown Source) ~[na:1.8.0_71]
    at org.terasology.persistence.internal.SaveTransactionHelper$2.postVisitDirectory(SaveTransactionHelper.java:103) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransactionHelper$2.postVisitDirectory(SaveTransactionHelper.java:72) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at java.nio.file.Files.walkFileTree(Unknown Source) ~[na:1.8.0_71]
    at java.nio.file.Files.walkFileTree(Unknown Source) ~[na:1.8.0_71]
    at org.terasology.persistence.internal.SaveTransactionHelper.mergeChanges(SaveTransactionHelper.java:72) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransaction.mergeChanges(SaveTransaction.java:465) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransaction.run(SaveTransaction.java:139) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.utilities.concurrency.TaskProcessor.run(TaskProcessor.java:49) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:1.8.0_71]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:1.8.0_71]
    at java.lang.Thread.run(Unknown Source) ~[na:1.8.0_71]
23:19:29.173 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
23:19:29.303 [main] ERROR o.terasology.engine.TerasologyEngine - Clean game shutdown after an uncaught exception failed
java.lang.RuntimeException: Saving failed
    at org.terasology.persistence.internal.ReadWriteStorageManager.checkSaveTransactionAndClearUpIfItIsDone(ReadWriteStorageManager.java:158) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.ReadWriteStorageManager.waitForCompletionOfPreviousSave(ReadWriteStorageManager.java:245) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.ReadWriteStorageManager.waitForCompletionOfPreviousSaveAndStartSaving(ReadWriteStorageManager.java:236) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:128) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:439) ~[engine-0.54.5.jar:62, origin/master, 62]
    at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:375) ~[engine-0.54.5.jar:62, origin/master, 62]
    at org.terasology.engine.Terasology.main(Terasology.java:154) [Terasology.jar:62, origin/master, 62]
Caused by: java.nio.file.DirectoryNotEmptyException: C:\Users\Marcel PC\Saved Games\Terasology\saves\Game 6\unmerged-changes
    at sun.nio.fs.WindowsFileSystemProvider.implDelete(Unknown Source) ~[na:1.8.0_71]
    at sun.nio.fs.AbstractFileSystemProvider.delete(Unknown Source) ~[na:1.8.0_71]
    at java.nio.file.Files.delete(Unknown Source) ~[na:1.8.0_71]
    at org.terasology.persistence.internal.SaveTransactionHelper$2.postVisitDirectory(SaveTransactionHelper.java:103) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransactionHelper$2.postVisitDirectory(SaveTransactionHelper.java:72) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at java.nio.file.Files.walkFileTree(Unknown Source) ~[na:1.8.0_71]
    at java.nio.file.Files.walkFileTree(Unknown Source) ~[na:1.8.0_71]
    at org.terasology.persistence.internal.SaveTransactionHelper.mergeChanges(SaveTransactionHelper.java:72) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransaction.mergeChanges(SaveTransaction.java:465) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.persistence.internal.SaveTransaction.run(SaveTransaction.java:139) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at org.terasology.utilities.concurrency.TaskProcessor.run(TaskProcessor.java:49) ~[engine-0.54.5.jar:62, master, 62, pre-alpha]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:1.8.0_71]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:1.8.0_71]
    at java.lang.Thread.run(Unknown Source) ~[na:1.8.0_71]
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30618344-rare-crash-attempting-to-auto-save-but-finding-a-non-empty-directory?utm_campaign=plugin&utm_content=tracker%2F5279&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5279&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/engine/src/main/java/org/terasology/persistence/internal/SaveTransactionHelper.java
+++ b/engine/src/main/java/org/terasology/persistence/internal/SaveTransactionHelper.java
@@ -20,6 +20,7 @@ import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.nio.file.AtomicMoveNotSupportedException;
+import java.nio.file.DirectoryNotEmptyException;
 import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -100,7 +101,23 @@ public class SaveTransactionHelper {
 
             @Override
             public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
-                Files.delete(dir);
+                try {
+                    Files.delete(dir);
+                }catch (DirectoryNotEmptyException e) {
+                    /**
+                     * Happens rarely for some players on windows (See issue #2160). Exact reason for this behavior is
+                     * unknown. Maybe they have some kind of background task that processes that creates a temporary
+                     * files in new directories to store some intermediate scan result.
+                     */
+                    logger.warn("The save job could not cleanup a temporarly created directory, it will retry once in one second");
+                    try {
+                        Thread.sleep(1000L);
+                    } catch (InterruptedException e1) {
+                        // Reset flag and ignore it
+                        Thread.currentThread().interrupt();
+                    }
+                    Files.delete(dir);
+                }
                 return FileVisitResult.CONTINUE;
             }
         });

